<launch>
    <arg name="initial_joint_positions"             default="-J waist 0 -J shoulder -1.5708 -J elbow 1.5708 -J forearm_roll -1.5708 -J wrist_angle -1.5708 -J wrist_rotate 3.14 -J left_finger 0 -J right_finger 0" doc="Initial joint configuration of the robot"/>
    <arg name="controllers"                         default="joint_state_controller cartesian_compliance_controller gripper_action_controller"/>
    <arg name="stopped_controllers"                 default="cartesian_motion_controller motion_control_handle cartesian_force_controller pos_joint_traj_controller joint_group_pos_controller"/>
    <arg name="controller_config_file"              default="$(find morpheus_moveit)/config/trossen_gazebo_controllers.yaml" />
    <arg name="moveit_controller_config_file"       default="$(find morpheus_moveit)/config/trossen_moveit_controllers.yaml" />
    <arg name="use_rviz"                            default="true" />

    <arg name="robot_model"                         default="vx300s"/>
    <arg name="robot_name"                          default="$(arg robot_model)"/>
    <arg name="dof"                                 default="6"/>

    <arg name="xacro_path"                          default="$(find morpheus_description)/urdf/scenes/trossen_node2_scene.xacro"/>
    <arg name="srdf_path"                           default="$(find interbotix_xsarm_moveit)/config/srdf/$(arg robot_model).srdf.xacro"/>

    <!-- Control motion and compliance controller with one handle -->
    <remap from="motion_control_handle/target_frame" to="target_frame" />
    <remap from="cartesian_motion_controller/target_frame" to="target_frame" />
    <remap from="cartesian_compliance_controller/target_frame" to="target_frame" />
    <!-- Control wrenches via one topic -->
    <remap from="cartesian_force_controller/target_wrench" to="target_wrench" />
    <remap from="cartesian_compliance_controller/target_wrench" to="target_wrench" />
    <!-- Read wrenches via one topic -->
    <remap from="cartesian_force_controller/ft_sensor_wrench" to="wrench_filtered" />
    <remap from="cartesian_compliance_controller/ft_sensor_wrench" to="wrench_filtered" />
    <!-- trossen driver -->
    <include file="$(find morpheus_teleop)/launch/trossen_gazebo.launch">
        <arg name="initial_joint_positions" value="$(arg initial_joint_positions)"/>
        <arg name="controllers" value="$(arg controllers)"/>
        <arg name="stopped_controllers" value="$(arg stopped_controllers)"/>
        <arg name="controller_config_file" value="$(arg controller_config_file)" />
        <arg name="robot_model" value="$(arg robot_model)" />
    </include>

    <!-- compliant trajectory action server -->
    <remap from="compliant_traj_action_server/target_wrench" to="target_wrench" />
    <remap from="compliant_traj_action_server/target_frame" to="target_frame" />
    <node 
        name="compliant_traj_action_server" 
        pkg="compliant_trajectory_control" 
        type="follow_compliant_trajectory_action_server" 
        output="screen"
    />

    <!-- launch moveit -->
    <include file="$(find morpheus_moveit)/launch/move_group.launch">
        <arg name="allow_trajectory_execution" value="true" />
        <arg name="moveit_controller_manager" value="simple" />
        <arg name="fake_execution_type" value="interpolate" />
        <arg name="info" value="true" />
        <arg name="debug" value="false" />
        <arg name="pipeline" value="ompl" />
        <arg name="load_robot_description" value="false" />
        <arg name="xacro_path" value="$(arg xacro_path)"/>
        <arg name="srdf_path" value="$(arg srdf_path)"/>
        <arg name="kinematics_path" value="$(find morpheus_teleop)/config/trossen_kinematics.yaml"/>
    </include>
    <!-- Load moveit controller list to the parameter server -->
    <rosparam file="$(arg moveit_controller_config_file)" ns="move_group"/>

    <!-- launch rviz -->
    <group if="$(arg use_rviz)">
        <arg name="my_rviz" value="-d $(find morpheus_moveit)/rviz/morpheus_gazebo.rviz" />
        <node name="rviz" pkg="rviz" type="rviz" respawn="false" args="$(arg my_rviz)"
            output="screen" />
    </group>

    <!-- launch xbox360_swap -->
    <include file="$(find morpheus_teleop)/launch/xbox360_swap_fake.launch">
        <arg name="twist_topic"                         value="twist_controller/command"/>
        <arg name="arm_group"                           value="interbotix_arm"/>
        <arg name="frame_id"                            value="vx300s/base_link"/>
        <arg name="end_effector"                        value="vx300s/ee_gripper_link"/>
    </include>

    <!-- launch key_to_joy -->
    <include file="$(find morpheus_teleop)/launch/key_to_joy.launch"/>

    <!-- launch trossen_joy -->
    <include file="$(find morpheus_teleop)/launch/trossen_joy.launch"/>
    
</launch>
